src/mame/sinclair/specnext_sprites.cpp

```diff
@@ -0,0 +1,282 @@
+    m_clip_window.set_origin(m_offset.first, m_offset.second);
+}
+
+void specnext_sprites_device::io_w(offs_t addr, u8 data)
+{
+    bool attr_num_change = 0;
+    if (addr == 0x5b)
+    {
+        // Ugh
+        gfx(0)->mark_all_dirty();
+        gfx(1)->mark_all_dirty();
+        gfx(2)->mark_all_dirty();
+        gfx(3)->mark_all_dirty();
+
+        m_sprite_pattern_ram[m_pattern_index] = data;
+        m_pattern_index++;
+    }
+    else if (addr == 0x303b)
+    {
+        m_pattern_index = (BIT(data, 0, TOTAL_PATTERN_BITS) << 8) | (data & 0x80);
+        m_attr_index = BIT(data, 0, TOTAL_SPRITES_BITS) << 3;
+        attr_num_change = 1;
+    }
+    else if (addr == 0x57)
+    {
+        m_sprites_cache.clear();
+        m_sprite_attr_ram[m_attr_index] = data;
+
+        const bool index_inc_attr_by_8 = BIT(m_attr_index, 2) || ((BIT(m_attr_index, 0, 3) == 0b011) && (BIT(data, 6) == 0));
+        const u16 index_inc_in_s = index_inc_attr_by_8
+            ? BIT(m_attr_index, 3, TOTAL_SPRITES_BITS)
+            : m_attr_index;
+        const u16 index_inc_out_s = index_inc_in_s + 1;
+
+        if (index_inc_attr_by_8 == 0)
+            m_attr_index = BIT(index_inc_out_s, 0, TOTAL_SPRITES_BITS + 3);
+        else
+        {
+            m_attr_index = BIT(index_inc_out_s, 0, TOTAL_SPRITES_BITS) << 3;
+            attr_num_change = 1;
+        }
+    }
+
+    if (attr_num_change && m_mirror_tie)
+    {
+        m_mirror_sprite_q &= ~0xff;
+        m_mirror_sprite_q |= BIT(m_attr_index, 3, TOTAL_SPRITES_BITS);
+        m_mirror_sprite_q |= m_pattern_index & 0x80;
+    }
+}
+
+void specnext_sprites_device::mirror_data_w(u8 mirror_data)
+{
+    bool mirror_num_change = 0;
+    if (m_mirror_index == 0b111)
+    {
+        m_mirror_sprite_q = mirror_data;
+        mirror_num_change = 1;
+    }
+    else if (m_mirror_inc) // ...
+    {
+        m_mirror_sprite_q &= ~0xff;
+        m_mirror_sprite_q |= (BIT(m_mirror_sprite_q, 0, TOTAL_SPRITES_BITS) + 1) & 0x7f;
+        m_mirror_sprite_q |= m_pattern_index & 0x80;
+        mirror_num_change = 1;
+    }
+
+    if (mirror_num_change && m_mirror_tie)
+    {
+        m_pattern_index = (BIT(m_mirror_sprite_q, 0, TOTAL_PATTERN_BITS) << 8) | (m_mirror_sprite_q & 0x80);
+        m_attr_index = BIT(m_mirror_sprite_q, 0, TOTAL_SPRITES_BITS) << 3;
+    }
+}
+
+void specnext_sprites_device::device_start()
+{
+    //save_item(NAME(m_offset));
+    save_item(NAME(m_zero_on_top));
+    save_item(NAME(m_border_clip_en));
+    save_item(NAME(m_over_border));
+    save_item(NAME(m_transp_colour));
+    save_item(NAME(m_clip_x1));
+    save_item(NAME(m_clip_x2));
+    save_item(NAME(m_clip_y1));
+    save_item(NAME(m_clip_y2));
+
+    save_item(NAME(m_mirror_tie));
+    save_item(NAME(m_mirror_index));
+    save_item(NAME(m_mirror_inc));
+
+    save_item(NAME(m_attr_index));
+    save_item(NAME(m_pattern_index));
+    save_item(NAME(m_mirror_sprite_q));
+}
```

Comment:
You’ll need to ensure the sprite cache is rebuilt after a saved state is restored.